src/main/java/me/oyurimatheus/nossoservicodepagamento/payment/purchase/OnlinePaymentAttemptsOrder.java

```diff
@@ -0,0 +1,43 @@
+package me.oyurimatheus.nossoservicodepagamento.payment.purchase;
+
+import me.oyurimatheus.nossoservicodepagamento.payment.purchase.gateways.PaymentGateway;
+import org.springframework.util.Assert;
+
+import java.util.Comparator;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+
+import static java.util.stream.Collectors.toList;
+
+class OnlinePaymentAttemptsOrder {
+
+    private List<PaymentGateway> gatewayOrder;
+
+    private OnlinePaymentAttemptsOrder(List<PaymentGateway> gatewayOrder) {
+        this.gatewayOrder = gatewayOrder;
+    }
+
+    public static OnlinePaymentAttemptsOrder makeGatewaysAttemptsOrderTo(Payment payment, Set<PaymentGateway> gateways) {
+
+        List<PaymentGateway> gatewayOrder = gateways.stream()
+                                                    .filter(gateway -> gateway.accept(payment))
```

Comment:
Eu to mais relaxado com aquele smell de consistência... tipo aqui você passa o payment no ```accept``` e depois passa o mesmo payment para o ```cost```. Existe a chance de trabalhar com payments diferentes, mas considero que é ok mesmo sabendo que tem risoc. Se eu tivesse implementando teria tentando seguido a ideia da consistência, mas já considero um exagero. 

Por outro lado você usa o payment como argumento para três métodos diferentes do mesmo gateway(2 aqui no método de fábrica e 1 no tryToPay) :) e isso sim é um smell... Pq você teria três métodos de instância recebendo o mesmo parâmetro? Tem cara que alguém deveria existir para guardar isso como atributo. 

E vou usar esse

Reply:
Como faria nesse caso sem receber duas vezes o pagamento no método `accept` e no `cost`? Como estavam no mesmo pipeline não vi muito problema também...

No outro caso, eu guardei o pagamento na classe e usei ele no `tryToPay`

Reply:
Opa, só para reforçar que seria um exagero da minha parte hauhaua. Mas eu fiz assim:

```
  gateways.stream()
   //retorna a Optional do objeto em si que faz a lógica de pagamento relacionado com o payment que é comparable //pelo cost
   .map(gateway -> gateway.accept(payment))
   .filter(Optional :: isPresent)
   .collect(Collectors.nao_lembro_o_nome_do_metodo(() -> new TreeSet()));
    
   .
```

Reply:
Eu pensei em usar um comparable haha, mas não lembro porque desisti dessa implementação. Mas pensei em algo assim no começo e delegava pro treeSet também haha